src/server/mod.rs

```diff
@@ -105,33 +112,49 @@ impl<S: SslServer> Server<HttpsListener<S>> {
 }
 
 
-impl<A: Accept> Server<A> where A::Output: Transport  {
+impl<A: Accept> Server<A> {
     /// Binds to a socket and starts handling connections.
     pub fn handle<H>(self, factory: H) -> ::Result<(Listening, ServerLoop<A, H>)>
     where H: HandlerFactory<A::Output> {
-        let addr = try!(self.listener.local_addr());
         let shutdown = Arc::new(AtomicBool::new(false));
-        let shutdown_rx = shutdown.clone();
-
+        
         let mut config = rotor::Config::new();
         config.slab_capacity(self.max_sockets);
         config.mio().notify_capacity(self.max_sockets);
         let keep_alive = self.keep_alive;
         let idle_timeout = self.idle_timeout;
         let mut loop_ = rotor::Loop::new(&config).unwrap();
+
+        let mut addrs = Vec::with_capacity(self.listeners.len());
+        let mut listeners = self.listeners.into_iter();
+
+        // Add the first listener. This one handles shutdown messages.
         let mut notifier = None;
         {
             let notifier = &mut notifier;
+            let listener = listeners.next().unwrap();
```

Comment:
I'm going to split listeners in to `lead_listener` and `other_listeners` to avoid having to unwrap here.